### PR TITLE
executor: fix a bug when join exists in subquery.

### LINKED ISSUE #2009: Got panic when do the following query
Please answer these questions before submitting your issue. Thanks!

1. What version of Go are you using (`go version`)?
1.7.3

2. What operating system and processor architecture are you using (`go env`)?
GOARCH="amd64"
GOBIN="/usr/local/go/bin"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/root/gopkg"
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build945606389=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"

3. What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.

```
drop table if exists t1, t2, t3;
create table t1 (a int NOT NULL, b int NOT NULL, key(a));
insert into t1 values 
  (0,0),(1,1),(2,2),(3,3),(4,4),(5,5),(6,6),(7,7),(8,8),(9,9);
create table t2 (a int NOT NULL, b int NOT NULL, key(a));
insert into t2 select * from t1;
update t2 set b=1;
create table t3 (a int, oref int);
insert into t3 values (1, 1), (NULL,1), (NULL,0);
select a, oref, 
       t3.a in (select t1.a from t1, t2 where t1.b=t2.a and t2.b=t3.oref) Z 
from t3;
drop table t1, t2, t3;
```

4. What did you expect to see?

5. What did you see instead?
TiDB returned the correct result, but got a panic and crashed after it.

### PR DESCRIPTION
**Pull request by hanfei1991**
When a join exitsts in subquery, the Close will be called before all rows are processed. In this case, we must ensure all channels have been closed.
fix #2009 
@shenli @coocood @zimulala @winoros @tiancaiamao @XuHuaiyu  PTAL

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -484,6 +486,9 @@ type hashJoinCtx struct {
 
 // Close implements the Executor Close interface.
 func (e *HashJoinExec) Close() error {
+	e.finished = true
```
Comment (tiancaiamao) [inline executor/executor.go]: I think you mean 

```
e.closeMu.Lock()
e.finished = true
e.closeMu.Unlock()
```

Comment (tiancaiamao) [inline executor/executor.go]: so what's the following two line for?

Comment (hanfei1991) [inline executor/executor.go]: add Lock for waiting waitJoinWorkersAndCloseResultChan unlock

Comment (zimulala) [inline executor/executor.go]: I think using a channel to notify is better. @hanfei1991 @tiancaiamao

Comment (hanfei1991) [inline executor/executor.go]: How to use a channel to implement it ? @zimulala

**Comment by hanfei1991** [comment]
@tiancaiamao @zimulala PTAL

File: executor/executor.go [hunk 2]
```diff
@@ -484,6 +487,10 @@ type hashJoinCtx struct {
 
 // Close implements the Executor Close interface.
 func (e *HashJoinExec) Close() error {
+	e.finished.Store(true)
+	select {
+	case _ = <-e.closeLock:
+	}
```
Comment (zimulala) [inline executor/executor.go]: Don't need to select and use `<-e.closeLock` directly.

File: executor/executor.go [hunk 3]
```diff
@@ -458,9 +459,11 @@ type HashJoinExec struct {
 	// targetTypes means the target the type that both smallHashKey and bigHashKey should convert to.
 	targetTypes []*types.FieldType
 
-	finished bool
-	// for sync multiple join workers.
+	finished atomic.Value
+	// For sync multiple join workers.
 	wg sync.WaitGroup
+	// closeMu add a lock for closing executor.
+	closeLock chan bool
```
Comment (zimulala) [inline executor/executor.go]: Using `struct {}` type  is better.

**Comment by zimulala** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**